{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5cff650e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcfpy\n",
    "import pandas as pd\n",
    "import os\n",
    "import sqlite3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1268b5db",
   "metadata": {},
   "outputs": [],
   "source": [
    "dbfile = os.path.join('exam.sqlite')\n",
    "db = sqlite3.connect(dbfile)\n",
    "vcf = '/lustre1/project/stg_00079/teaching/data/snps.annotated.vcf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "bdf2a4a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dbfile : exam.sqlite\n",
      "db     : <sqlite3.Connection object at 0x2add5f7a0d40>\n",
      "vcf    : /lustre1/project/stg_00079/teaching/data/snps.annotated.vcf\n"
     ]
    }
   ],
   "source": [
    "print(f\"dbfile : {dbfile}\")\n",
    "print(f\"db     : {db}\")\n",
    "print(f\"vcf    : {vcf}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4ec1109f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3505\n",
      "7010\n",
      "19640\n"
     ]
    }
   ],
   "source": [
    "# Empty lists to store the data in\n",
    "snp_records = []\n",
    "effect_records = []\n",
    "call_records = []\n",
    "\n",
    "# These are the columns from a SNPeff record - easy to later assign column names\n",
    "effect_rec_names = \"\"\"snp allele effect impact gene gene_id feature_type feature_id \n",
    "                      biotype rank hgvs.c hgvs.p cdna_pos cds_pos prot_pos distance_to_feature\n",
    "                      messages\"\"\".split()\n",
    "\n",
    "# start parsing through the VCF:\n",
    "j = 0\n",
    "\n",
    "#open the vcf iterator:\n",
    "reader = vcfpy.Reader.from_path(vcf)\n",
    "\n",
    "for i, record in enumerate(reader):\n",
    "    j += 1\n",
    "    \n",
    "    # we used vt decompose - so I expect NO multiallelic SNPs\n",
    "    assert len(record.ALT) == 1 \n",
    "    \n",
    "    # this is the ALT allele (first and only)\n",
    "    alt = record.ALT[0]\n",
    "    \n",
    "    # compose a SNP name for joins later on\n",
    "    snp_name = f\"{record.CHROM}:{record.POS}:{record.REF}:{alt.value}\"\n",
    "    \n",
    "    # Store a SNP record - I collect everything now in a list of dictionaries\n",
    "    # that I will later convert to a Pandas dataframe\n",
    "    snp_records.append(\n",
    "        dict(snp = snp_name,\n",
    "             chrom = record.CHROM,\n",
    "             pos = record.POS,\n",
    "             quality = record.QUAL,\n",
    "             ref = record.REF,\n",
    "             type = alt.type,\n",
    "             alt = alt.value))\n",
    "    \n",
    "    \n",
    "    # Now I will parse through the individual calls\n",
    "    for call_record in record.calls:\n",
    "        \n",
    "        #name of the sample (TLE66_N or TLE66_T)\n",
    "        sample = os.path.basename(call_record.sample)\n",
    "        sample = sample.replace('.bam', '')\n",
    "\n",
    "        # simple genotype - -1 means not called, otherwise it's the number of ALT alleles observer\n",
    "        # this helps in later queries \n",
    "        # Note - this is not perfect - it does ignore uncalled loci (./.)\n",
    "        genotype_simple = call_record.data['GT'].count('1')\n",
    "        \n",
    "        call_records.append(\n",
    "            dict(snp = snp_name,\n",
    "                 sample = sample,\n",
    "                 genotype = call_record.data['GT'], \n",
    "                 genotype_simple = genotype_simple))\n",
    "\n",
    "    # and parse through all snpEff annotations\n",
    "    for ann in record.INFO['ANN']:\n",
    "        ann = ann.split('|')\n",
    "        # create a dictionary of all fields\n",
    "        eff_record = dict(zip(effect_rec_names, [snp_name] + ann))\n",
    "        #from pprint import pprint\n",
    "        #pprint(eff_record)\n",
    "        # convert distance to feature to integer\n",
    "        try:\n",
    "            eff_record['distance_to_feature'] = int(eff_record.get(intfield))\n",
    "        except:\n",
    "            eff_record['distance_to_feature'] = -1\n",
    "            \n",
    "        effect_records.append(eff_record)\n",
    "\n",
    "print(len(snp_records))\n",
    "print(len(call_records))\n",
    "print(len(effect_records))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "02d6b2c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "snp records 3505\n",
      "call records 7010\n",
      "effect records 19640\n"
     ]
    }
   ],
   "source": [
    "#convert lists of dicts to a DataFrame\n",
    "snp_records = pd.DataFrame.from_records(snp_records)\n",
    "call_records = pd.DataFrame.from_records(call_records)\n",
    "effect_records = pd.DataFrame.from_records(effect_records)\n",
    "\n",
    "#Save to disk\n",
    "print('snp records', snp_records.to_sql('snp', db, if_exists='replace', index=False))\n",
    "print('call records', call_records.to_sql('snp_call', db, if_exists='replace', index=False))\n",
    "print('effect records', effect_records.to_sql('snp_effect', db, if_exists='replace', index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7a05f02b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>quality</th>\n",
       "      <th>ref</th>\n",
       "      <th>type</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9:127578816:C:T</td>\n",
       "      <td>chr9</td>\n",
       "      <td>127578816</td>\n",
       "      <td>36.1174</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9:127578974:A:G</td>\n",
       "      <td>chr9</td>\n",
       "      <td>127578974</td>\n",
       "      <td>422.7380</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9:127579080:A:G</td>\n",
       "      <td>chr9</td>\n",
       "      <td>127579080</td>\n",
       "      <td>172.0220</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9:127663498:C:T</td>\n",
       "      <td>chr9</td>\n",
       "      <td>127663498</td>\n",
       "      <td>66.0472</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9:127674824:G:T</td>\n",
       "      <td>chr9</td>\n",
       "      <td>127674824</td>\n",
       "      <td>24.6981</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  snp chrom        pos   quality ref type alt\n",
       "0  chr9:127578816:C:T  chr9  127578816   36.1174   C  SNV   T\n",
       "1  chr9:127578974:A:G  chr9  127578974  422.7380   A  SNV   G\n",
       "2  chr9:127579080:A:G  chr9  127579080  172.0220   A  SNV   G\n",
       "3  chr9:127663498:C:T  chr9  127663498   66.0472   C  SNV   T\n",
       "4  chr9:127674824:G:T  chr9  127674824   24.6981   G  SNV   T"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql(\"SELECT * FROM snp LIMIT 5\", db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "69faffbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>sample</th>\n",
       "      <th>genotype</th>\n",
       "      <th>genotype_simple</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9:127578816:C:T</td>\n",
       "      <td>TLE66_N</td>\n",
       "      <td>0/1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9:127578816:C:T</td>\n",
       "      <td>TLE66_T</td>\n",
       "      <td>0/1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9:127578974:A:G</td>\n",
       "      <td>TLE66_N</td>\n",
       "      <td>0/1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9:127578974:A:G</td>\n",
       "      <td>TLE66_T</td>\n",
       "      <td>0/1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9:127579080:A:G</td>\n",
       "      <td>TLE66_N</td>\n",
       "      <td>0/1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  snp   sample genotype  genotype_simple\n",
       "0  chr9:127578816:C:T  TLE66_N      0/1                1\n",
       "1  chr9:127578816:C:T  TLE66_T      0/1                1\n",
       "2  chr9:127578974:A:G  TLE66_N      0/1                1\n",
       "3  chr9:127578974:A:G  TLE66_T      0/1                1\n",
       "4  chr9:127579080:A:G  TLE66_N      0/1                1"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql(\"SELECT * FROM snp_call LIMIT 5\", db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d51df9f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>allele</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>biotype</th>\n",
       "      <th>rank</th>\n",
       "      <th>hgvs.c</th>\n",
       "      <th>hgvs.p</th>\n",
       "      <th>cdna_pos</th>\n",
       "      <th>cds_pos</th>\n",
       "      <th>prot_pos</th>\n",
       "      <th>distance_to_feature</th>\n",
       "      <th>messages</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9:127578816:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>ENSG00000136854</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000637521.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td></td>\n",
       "      <td>c.-724C&gt;T</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>-1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9:127578816:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>NIBAN2</td>\n",
       "      <td>ENSG00000136830</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000373314.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1/13</td>\n",
       "      <td>c.16+106G&gt;A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>-1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9:127578974:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>NIBAN2</td>\n",
       "      <td>ENSG00000136830</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000373314.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1/14</td>\n",
       "      <td>c.-37T&gt;C</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>-1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9:127578974:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>ENSG00000136854</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000637521.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td></td>\n",
       "      <td>c.-566A&gt;G</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>-1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9:127579080:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>NIBAN2</td>\n",
       "      <td>ENSG00000136830</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000373314.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td></td>\n",
       "      <td>c.-143T&gt;C</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>-1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  snp allele                 effect    impact    gene  \\\n",
       "0  chr9:127578816:C:T      T  upstream_gene_variant  MODIFIER  STXBP1   \n",
       "1  chr9:127578816:C:T      T         intron_variant  MODIFIER  NIBAN2   \n",
       "2  chr9:127578974:A:G      G    5_prime_UTR_variant  MODIFIER  NIBAN2   \n",
       "3  chr9:127578974:A:G      G  upstream_gene_variant  MODIFIER  STXBP1   \n",
       "4  chr9:127579080:A:G      G  upstream_gene_variant  MODIFIER  NIBAN2   \n",
       "\n",
       "           gene_id feature_type         feature_id         biotype  rank  \\\n",
       "0  ENSG00000136854   transcript  ENST00000637521.2  protein_coding         \n",
       "1  ENSG00000136830   transcript  ENST00000373314.7  protein_coding  1/13   \n",
       "2  ENSG00000136830   transcript  ENST00000373314.7  protein_coding  1/14   \n",
       "3  ENSG00000136854   transcript  ENST00000637521.2  protein_coding         \n",
       "4  ENSG00000136830   transcript  ENST00000373314.7  protein_coding         \n",
       "\n",
       "        hgvs.c hgvs.p cdna_pos cds_pos prot_pos  distance_to_feature messages  \n",
       "0    c.-724C>T                                                    -1           \n",
       "1  c.16+106G>A                                                    -1           \n",
       "2     c.-37T>C                                                    -1           \n",
       "3    c.-566A>G                                                    -1           \n",
       "4    c.-143T>C                                                    -1           "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# .T transposes - easier viewing\n",
    "pd.read_sql(\"SELECT * FROM snp_effect LIMIT 5\", db)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55061585",
   "metadata": {},
   "source": [
    "#### Find high impact variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "84ee65fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>gene</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9:128632018:A:G</td>\n",
       "      <td>SPTAN1</td>\n",
       "      <td>splice_acceptor_variant&amp;intron_variant</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9:133257521:T:TC</td>\n",
       "      <td>ABO</td>\n",
       "      <td>splice_acceptor_variant&amp;splice_donor_variant&amp;i...</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9:135077073:G:GAA</td>\n",
       "      <td>OLFM1</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9:135077150:GCA:G</td>\n",
       "      <td>OLFM1</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9:137245305:G:A</td>\n",
       "      <td>FAM166A</td>\n",
       "      <td>splice_donor_variant&amp;intron_variant</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr9:138046919:C:A</td>\n",
       "      <td>CACNA1B</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    snp     gene  \\\n",
       "0    chr9:128632018:A:G   SPTAN1   \n",
       "1   chr9:133257521:T:TC      ABO   \n",
       "2  chr9:135077073:G:GAA    OLFM1   \n",
       "3  chr9:135077150:GCA:G    OLFM1   \n",
       "4    chr9:137245305:G:A  FAM166A   \n",
       "5    chr9:138046919:C:A  CACNA1B   \n",
       "\n",
       "                                              effect impact  \n",
       "0             splice_acceptor_variant&intron_variant   HIGH  \n",
       "1  splice_acceptor_variant&splice_donor_variant&i...   HIGH  \n",
       "2                                 frameshift_variant   HIGH  \n",
       "3                                 frameshift_variant   HIGH  \n",
       "4                splice_donor_variant&intron_variant   HIGH  \n",
       "5                                        stop_gained   HIGH  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql(\"\"\"\n",
    "    SELECT DISTINCT snp.snp, snp_effect.gene, \n",
    "                    snp_effect.effect, snp_effect.impact\n",
    "               FROM snp, snp_effect\n",
    "              WHERE snp.snp = snp_effect.snp\n",
    "                AND snp_effect.impact == 'HIGH'\n",
    "            \"\"\", db)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75a10c52",
   "metadata": {},
   "source": [
    "#### Find MODERATE and HIGH impact variants that differ between normal and tumor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "14541377",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>genotype_normal</th>\n",
       "      <th>genotype_tumor</th>\n",
       "      <th>gene</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9:127802988:A:G</td>\n",
       "      <td>1/1</td>\n",
       "      <td>./.</td>\n",
       "      <td>FPGS</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9:130396267:G:A</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>HMCN2</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9:130408923:G:A</td>\n",
       "      <td>1/1</td>\n",
       "      <td>0/1</td>\n",
       "      <td>HMCN2</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9:131475071:C:T</td>\n",
       "      <td>./.</td>\n",
       "      <td>1/1</td>\n",
       "      <td>PRRC2B</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9:131475956:C:T</td>\n",
       "      <td>0/1</td>\n",
       "      <td>1/1</td>\n",
       "      <td>PRRC2B</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr9:135077073:G:GAA</td>\n",
       "      <td>./.</td>\n",
       "      <td>1/1</td>\n",
       "      <td>OLFM1</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr9:135077150:GCA:G</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/1</td>\n",
       "      <td>OLFM1</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr9:136463081:C:T</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0/0</td>\n",
       "      <td>SEC16A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr9:136503310:A:T</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/1</td>\n",
       "      <td>NOTCH1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr9:136687198:C:T</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/1</td>\n",
       "      <td>AGPAT2</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr9:136837879:G:C</td>\n",
       "      <td>1/1</td>\n",
       "      <td>./.</td>\n",
       "      <td>RABL6</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr9:137252821:C:T</td>\n",
       "      <td>1/1</td>\n",
       "      <td>./.</td>\n",
       "      <td>STPG3</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr9:138046919:C:A</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/1</td>\n",
       "      <td>CACNA1B</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     snp genotype_normal genotype_tumor     gene  \\\n",
       "0     chr9:127802988:A:G             1/1            ./.     FPGS   \n",
       "1     chr9:130396267:G:A             0/1            0/0    HMCN2   \n",
       "2     chr9:130408923:G:A             1/1            0/1    HMCN2   \n",
       "3     chr9:131475071:C:T             ./.            1/1   PRRC2B   \n",
       "4     chr9:131475956:C:T             0/1            1/1   PRRC2B   \n",
       "5   chr9:135077073:G:GAA             ./.            1/1    OLFM1   \n",
       "6   chr9:135077150:GCA:G             0/0            0/1    OLFM1   \n",
       "7     chr9:136463081:C:T             0/1            0/0   SEC16A   \n",
       "8     chr9:136503310:A:T             0/0            0/1   NOTCH1   \n",
       "9     chr9:136687198:C:T             0/0            0/1   AGPAT2   \n",
       "10    chr9:136837879:G:C             1/1            ./.    RABL6   \n",
       "11    chr9:137252821:C:T             1/1            ./.    STPG3   \n",
       "12    chr9:138046919:C:A             0/0            0/1  CACNA1B   \n",
       "\n",
       "                effect    impact  \n",
       "0     missense_variant  MODERATE  \n",
       "1     missense_variant  MODERATE  \n",
       "2     missense_variant  MODERATE  \n",
       "3     missense_variant  MODERATE  \n",
       "4     missense_variant  MODERATE  \n",
       "5   frameshift_variant      HIGH  \n",
       "6   frameshift_variant      HIGH  \n",
       "7     missense_variant  MODERATE  \n",
       "8     missense_variant  MODERATE  \n",
       "9     missense_variant  MODERATE  \n",
       "10    missense_variant  MODERATE  \n",
       "11    missense_variant  MODERATE  \n",
       "12         stop_gained      HIGH  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql(\"\"\"SELECT DISTINCT snp.snp, \n",
    "                      sc1.genotype as genotype_normal, \n",
    "                      sc2.genotype as genotype_tumor,\n",
    "                      snp_effect.gene, snp_effect.effect, snp_effect.impact\n",
    "                 FROM snp, snp_call as sc1, snp_call as sc2, snp_effect\n",
    "                WHERE snp.snp = sc1.snp\n",
    "                  AND snp.snp = sc2.snp\n",
    "                  AND sc1.sample = 'TLE66_N'\n",
    "                  AND sc2.sample = 'TLE66_T'\n",
    "                  AND sc1.genotype != sc2.genotype\n",
    "                  AND snp.snp = snp_effect.snp\n",
    "                  AND ( snp_effect.impact = 'MODERATE'\n",
    "                        OR snp_effect.impact = 'HIGH' )\n",
    "            \"\"\", db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a88749a4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "large_omics",
   "language": "python",
   "name": "large_omics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
